File: app/src/main/java/com/example/bloodbank/ChatActivity.java
package com.example.bloodbank;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.example.bloodbank.Adapter.MessageAdapter;
import com.example.bloodbank.Model.Message;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class ChatActivity extends AppCompatActivity {

    private CircleImageView profileImage;
    private TextView username;
    private RecyclerView recyclerView;
    private EditText messageInput;
    private ImageButton sendButton;
    private Toolbar toolbar;

    private MessageAdapter messageAdapter;
    private List<Message> messageList;
    private String userId;
    private String profilePic;
    private String userName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        // Get user info from intent
        userId = getIntent().getStringExtra("userid");
        profilePic = getIntent().getStringExtra("profilepic");
        userName = getIntent().getStringExtra("username");

        // Initialize views
        initializeViews();
        
        // Setup toolbar
        setupToolbar();
        
        // Setup RecyclerView
        setupRecyclerView();
        
        // Load messages
        loadMessages();
        
        // Setup send button
        sendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String message = messageInput.getText().toString().trim();
                if (!message.isEmpty()) {
                    sendMessage(message);
                    messageInput.setText("");
                }
            }
        });
    }
    
    private void initializeViews() {
        profileImage = findViewById(R.id.profileImage);
        username = findViewById(R.id.username);
        recyclerView = findViewById(R.id.recyclerView);
        messageInput = findViewById(R.id.messageInput);
        sendButton = findViewById(R.id.sendButton);
        toolbar = findViewById(R.id.toolbar);
    }
    
    private void setupToolbar() {
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }
        
        username.setText(userName);
        if (profilePic != null && !profilePic.isEmpty()) {
            // Load profile image using Glide
            com.bumptech.glide.Glide.with(this)
                    .load(profilePic)
                    .placeholder(R.drawable.profile_pic)
                    .into(profileImage);
        } else {
            profileImage.setImageResource(R.drawable.profile_pic);
        }
    }
    
    private void setupRecyclerView() {
        messageList = new ArrayList<>();
        messageAdapter = new MessageAdapter(this, messageList);
        
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setStackFromEnd(true);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(messageAdapter);
    }
    
    private void loadMessages() {
        DatabaseReference chatRef = FirebaseDatabase.getInstance().getReference().child("Chats");
        chatRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                messageList.clear();
                for (DataSnapshot dataSnapshot : snapshot.getChildren()) {
                    Message message = dataSnapshot.getValue(Message.class);
                    if (message != null) {
                        if (message.getSenderId().equals(FirebaseAuth.getInstance().getCurrentUser().getUid()) &&
                                message.getReceiverId().equals(userId) ||
                                message.getSenderId().equals(userId) &&
                                        message.getReceiverId().equals(FirebaseAuth.getInstance().getCurrentUser().getUid())) {
                            messageList.add(message);
                        }
                    }
                }
                messageAdapter.notifyDataSetChanged();
                recyclerView.scrollToPosition(messageList.size() - 1);
            }
            
            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                // Handle error silently
            }
        });
    }
    
    private void sendMessage(String messageText) {
        DatabaseReference chatRef = FirebaseDatabase.getInstance().getReference().child("Chats");
        String pushId = chatRef.push().getKey();
        
        if (pushId != null) {
            HashMap<String, Object> messageMap = new HashMap<>();
            messageMap.put("senderId", FirebaseAuth.getInstance().getCurrentUser().getUid());
            messageMap.put("receiverId", userId);
            messageMap.put("message", messageText);
            messageMap.put("timestamp", System.currentTimeMillis());
            
            chatRef.child(pushId).setValue(messageMap);
        }
    }
    
    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
} 

File: app/src/main/java/com/example/bloodbank/ChatActivity.kt
package com.example.bloodbank

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.bloodbank.Adapter.MessageAdapter
import dagger.hilt.android.AndroidEntryPoint
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.coroutines.launch

/**
 * ChatActivity - Displays chat conversation between two users.
 * Follows MVVM architecture pattern with ChatViewModel.
 */
@AndroidEntryPoint
class ChatActivity : AppCompatActivity() {
    
    private val viewModel: ChatViewModel by viewModels()
    
    // Views
    private lateinit var profileImage: CircleImageView
    private lateinit var username: TextView
    private lateinit var recyclerView: RecyclerView
    private lateinit var messageInput: EditText
    private lateinit var sendButton: ImageButton
    private lateinit var toolbar: Toolbar
    
    // Adapter
    private lateinit var messageAdapter: MessageAdapter
    
    // User info from intent
    private var userId: String = ""
    private var profilePic: String = ""
    private var userName: String = ""
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        
        // Get user info from intent
        userId = intent.getStringExtra("userid") ?: ""
        profilePic = intent.getStringExtra("profilepic") ?: ""
        userName = intent.getStringExtra("username") ?: ""
        
        // Initialize views
        initializeViews()
        
        // Setup toolbar
        setupToolbar()
        
        // Setup RecyclerView
        setupRecyclerView()
        
        // Setup send button
        setupSendButton()
        
        // Load messages
        viewModel.loadMessages(userId)
        
        // Observe ViewModel state
        observeViewModel()
    }
    
    private fun initializeViews() {
        profileImage = findViewById(R.id.profileImage)
        username = findViewById(R.id.username)
        recyclerView = findViewById(R.id.recyclerView)
        messageInput = findViewById(R.id.messageInput)
        sendButton = findViewById(R.id.sendButton)
        toolbar = findViewById(R.id.toolbar)
    }
    
    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = ""
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
        
        username.text = userName
        
        if (profilePic.isNotEmpty()) {
            Glide.with(this)
                .load(profilePic)
                .placeholder(R.drawable.profile_pic)
                .into(profileImage)
        } else {
            profileImage.setImageResource(R.drawable.profile_pic)
        }
    }
    
    private fun setupRecyclerView() {
        messageAdapter = MessageAdapter(this, mutableListOf())
        
        recyclerView.apply {
            layoutManager = LinearLayoutManager(this@ChatActivity).apply {
                stackFromEnd = true
            }
            adapter = messageAdapter
        }
    }
    
    private fun setupSendButton() {
        sendButton.setOnClickListener {
            val message = messageInput.text.toString().trim()
            if (message.isNotEmpty()) {
                viewModel.sendMessage(message)
                messageInput.setText("")
            }
        }
    }
    
    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                // Observe messages
                launch {
                    viewModel.messages.collect { messages ->
                        messageAdapter.updateMessages(messages)
                        if (messages.isNotEmpty()) {
                            recyclerView.scrollToPosition(messages.size - 1)
                        }
                    }
                }
                
                // Observe errors
                launch {
                    viewModel.error.collect { error ->
                        error?.let {
                            Toast.makeText(this@ChatActivity, it, Toast.LENGTH_SHORT).show()
                            viewModel.clearError()
                        }
                    }
                }
                
                // Observe loading state (optional - can show progress indicator)
                launch {
                    viewModel.isLoading.collect { isLoading ->
                        // Optional: show/hide loading indicator
                    }
                }
            }
        }
    }
    
    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }
}
